rclcpp/src/rclcpp/executors/static_executor_waitable.cpp

```diff
@@ -0,0 +1,264 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/executors/static_single_threaded_executor.hpp"
+#include "rclcpp/executors/static_executor_waitable.hpp"
+
+using rclcpp::executors::StaticExecutorWaitable;
+
+StaticExecutorWaitable::~StaticExecutorWaitable()
+{
+  // Disassociate all nodes
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (node) {
+      std::atomic_bool & has_executor = node->get_associated_with_executor_atomic();
+      has_executor.store(false);
+    }
+  }
+  exec_list_.clear();
+  weak_nodes_.clear();
+  guard_conditions_.clear();
+}
+
+void
+StaticExecutorWaitable::init(
+       rcl_wait_set_t* p_wait_set,
+       memory_strategy::MemoryStrategy::SharedPtr& memory_strategy,
+       StaticExecutorWaitable::SharedPtr this_shared_ptr)
+{
+  // Empty initialize executable list
+  executor::ExecutableList exec_list_();
+  // Get executor's wait_set_ pointer
+  p_wait_set_ = p_wait_set;
+  // Get executor's memory strategy ptr
+  if (memory_strategy == nullptr) {
+    throw std::runtime_error("Received NULL memory strategy in executor waitable.");
+  }
+  memory_strategy_ = memory_strategy;
+  // Get this StaticExecutorWaitable shared pointer
+  this_shared_ptr_ = this_shared_ptr;
```

Comment:
this looks very bad

Reply:
I couldn't find a workaroud. The waitable member function `execute()`  has to be able from inside, to regenerate the executable list and add a waitable object (itself). I can't do this from outside, unless I call the `StaticSingleThreadedExecutor` to do it, which would introduce a circular dependency. 
As for deleting the shared pointer, it should be deleted on the destructor of the class.

Reply:
Maybe if I could on the fly generate a shared ptr based on `this` keyword I could manage to avoid `this_shared_ptr`

Reply:
There is exactly a C++ thing called `shared_from_this`
https://en.cppreference.com/w/cpp/memory/enable_shared_from_this

however, we should not use it here